dbcon/mysql/rbo_apply_parallel_ces.cpp

```diff
@@ -134,127 +172,263 @@ execplan::SimpleColumn* findSuitableKeyColumn(execplan::CalpontSelectExecutionPl
     bounds.push_back({currentLowerBound, currentUpperBound});
   }
-
-  // Add last range
-  // NB despite the fact that currently Histogram_json_hb has the last bucket that has end as its start
-  auto lastBucket = columnStatistics.get_json_histogram().begin() + (numberOfUnionUnits - 1) * numberOfBucketsPerUnionUnit;
-  uint64_t currentLowerBound = *(uint32_t*)lastBucket->start_value.data();
-  uint64_t currentUpperBound = *(uint32_t*)columnStatistics.get_last_bucket_end_endp().data();
-  bounds.push_back({currentLowerBound, currentUpperBound});
+  for (auto& bucket : columnStatistics.get_json_histogram())
+  {
+    T currentLowerBound = *(uint32_t*)bucket.start_value.data();
+    std::cout << "Bucket: " << currentLowerBound << std::endl;
+  }
+  // auto penultimateBucket = columnStatistics.get_json_histogram().begin() + numberOfUnionUnits * numberOfBucketsPerUnionUnit;
+  // T currentLowerBound = *(uint32_t*)penultimateBucket->start_value.data();
+  // T currentUpperBound = *(uint32_t*)columnStatistics.get_last_bucket_end_endp().data();
+  // bounds.push_back({currentLowerBound, currentUpperBound});
 
   for (auto& bound : bounds)
   {
-    auto clonedCSEP = csep.cloneWORecursiveSelects();
-    // Add BETWEEN based on key column range
-    clonedCSEP->filters(filtersWithNewRangeAddedIfNeeded(clonedCSEP, *keyColumn, bound));
-    unionVec.push_back(clonedCSEP);
+    std::cout << "Bound: " << bound.first << " " << bound.second << std::endl;
+  }
+
+  return bounds;
+}
+
+// TODO char and other numerical types support
+execplan::CalpontSelectExecutionPlan::SelectList makeUnionFromTable(
+    execplan::CalpontSelectExecutionPlan& csep, execplan::CalpontSystemCatalog::TableAliasName& table,
+    optimizer::RBOptimizerContext& ctx)
+{
+  execplan::CalpontSelectExecutionPlan::SelectList unionVec;
+
+  // SC type controls an integral type used to produce suitable filters. The continuation of this function
+  // should become a template function based on SC type.
+  execplan::SimpleColumn* keyColumn = findSuitableKeyColumn(csep, table, ctx);
+  if (!keyColumn)
+  {
+    return unionVec;
+  }
+
+  // TODO char and other numerical types support
+  auto boundsOpt = populateRangeBounds<uint64_t>(keyColumn, ctx);
+  if (!boundsOpt.has_value())
+  {
+    return unionVec;
+  }
+
+  auto& bounds = boundsOpt.value();
+
+  // These bounds produce low <= col < high
+  if (bounds.size() > 1)
+  {
+    for (size_t i = 0; i <= bounds.size() - 2; ++i)
+    {
+      auto clonedCSEP = csep.cloneForTableWORecursiveSelects(table);
+      // Add BETWEEN based on key column range
+      clonedCSEP->filters(filtersWithNewRange(clonedCSEP, *keyColumn, bounds[i], false));
+      unionVec.push_back(clonedCSEP);
+    }
   }
+  // This last bound produces low <= col <= high
+  auto clonedCSEP = csep.cloneForTableWORecursiveSelects(table);
+  clonedCSEP->filters(filtersWithNewRange(clonedCSEP, *keyColumn, bounds.back(), true));
+  unionVec.push_back(clonedCSEP);
```

Comment:
```suggestion
    for (size_t i = 0; i < bounds.size(); ++i)
    {
      auto clonedCSEP = csep.cloneForTableWORecursiveSelects(table);
      // Add BETWEEN based on key column range
      bool isLast = i == bounds.size() - 1;
      clonedCSEP->filters(filtersWithNewRange(clonedCSEP, *keyColumn, bounds[i], isLast));
      unionVec.push_back(clonedCSEP);
    }
  }
```

Reply:
There will be an additional logic, so that a fast bool computation won't safe the day. For the last  bound I have to take a special attribute from Histogram_json_hb.